add os testing docs (#2878)
balena-os/meta-balena: tests/README.md

# Balena OS automated testing

> Balena OS is tested on individual device types before being released to production. This folder contains these tests. 

Reviewer: ```suggestion
BalenaOS is tested on individual device types before being released to production. This directory contains these tests. 
```

The tests are split into 3 test suites:

> - os: this test suite verifies that an unmanaged balena OS works as intended. 

Reviewer: ```suggestion
- os: this test suite verifies that an unmanaged balenaOS works as intended. 
```

> - cloud: this test suite verifies that a balena OS version integrates correctly with balena cloud features

Reviewer: ```suggestion
- cloud: this test suite verifies that a balenaOS version integrates correctly with balena cloud features
```

> - hup: this test suite verifies that host app update (HUP) features of balena OS work correctly

Reviewer: ```suggestion
- hup: this test suite verifies that host app update (HUP) features of balenaOS work correctly
```

Each folder contains a readme with more details about each individual test suite

## How are the tests ran

> On each PR in meta-balena, jenkins yocto jobs create a "draft" OS image from the contents of the branch in that PR. 

Reviewer: ```suggestion
On each PR in meta-balena, Jenkins `yocto` jobs create a "draft" OS image from the contents of the branch in that PR. 
```

The test suites are then run against virtualised QEMU devices, for `x86`, `aarch64`, and `amd64` architectures.

> Once a PR is merged, a PR is automatically made on the device repos, bumping the meta-balena version. 

Reviewer: ```suggestion
Once a meta-balena PR is merged, a wave of new PRs are automatically opened on the device repos, bumping the meta-balena version. 
```

> The same workflow is then executed on the physical hardware corresponding to that device type. 

Reviewer: ```suggestion
The same test suites are then executed on the physical hardware corresponding to that device type. 
```

## Testing framework

We use our testing framework to run these tests, called [Leviathan](https://github.com/balena-os/leviathan).
This framework allows us to run the same tests on all device types, and on both virtualised and physical environments.
It provides a common layer of abstraction that is shared by both the physical device setup, and the virtual.  

All tests in the `suites/<SUITE_NAME>` directories are executed by the Leviathan framework.

### Testing on virtualised devices

Leviathan simplifies the setup of a virtual device, running via QEMU. 

### Testing on physical devices

In order to run tests of new OS versions on real hardware, we needed a tool to automated the provisioning of these devices, and to be able to automatically interact with its interfaces. To this end, we developed a tool called the [AutoKit](https://github.com/balena-io-hardware/autokit).
It allows for:

- multiplexing SD cards between a USB host and the device under test (DUT)
- controlling the power to the DUT
- capturing video output
- providing a serial terminal
- providing a wifi AP or an ethernet connection to the DUT
- bluetooth
- a relay that can automate toggling of a jumper or switch

> This tool is composed of entirely CotS hardware, and all software is open source.

Reviewer: What is CotS (for me and the doc)?
